### ISSUE TITLE
How to use mse loss

### ISSUE THREAD

**Issue by ifzhang** [xingyizhou/CenterNet#718]
Hi, thanks for your nice work! Have you ever evaluated how much focal loss is better than mse loss? I use '--mse_loss' and it says there's no opt.hm_gauss. I set the radius 3 and find the results are very bad. Could you tell me what value of radius I should set if using mse loss? Thanks!

**Comment by ifzhang** [comment]
@xingyizhou

**Comment by xingyizhou** [comment]
Thank you for your question. I am very interested in this question too. I have played with MSE loss for a week, but never made it work. Here are the things I have considered: 

- How to normalize the loss. If it is `((pred - gt)**2).mean()` as is in single person human pose literature, it will be very small (most cells are 0). I also tried something like (could not found the exact log) `(pred - gt).sum() / B / H / W` or `(pred - gt).sum() / B / C` or `(pred - gt).sum() / N`, but none of them gives >10 mAP on COCO. Here B is batchsize, C is the # categories, and N is the # positive samples. I did these experiments on Feb 2019 before ICCV19 deadline, so some settings might be outdated in the current codebase.

- How to apply focal loss. First of all many variants of focal loss work in the current sigmoid cross-entropy form. E.g., only apply focal loss on negative cells, or not use (1-M_{xy})^\alpha, just masking out the losses when the loss of a cell is smaller than 1e-4. I haven't tried them for MSE.

For your question, I think the radius value from the single person human pose community is input_size / 128. So for a 512x512 input it should be 4. I feel the above two issues are more essential than the radius.

A side suggestion is that don't use this codebase. Use CenterTrack codebase. It supports full functionality of CenterNet, and has been refactored a few times. The focal loss implementation there is more clear. I planed to move the code to this repo but still haven't finished all the consistent testing ...
